### PR TITLE
ref impl: Clean up sequencer

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**
This better handles the fork choice update in the rollup node, exposes the fork choice method throughout,
cleans up the L2 Engine API interface, and random aspects of block creation. 

The `l2.Source` is the Ethereum client wrapper for the L2 execution engine. It directly wraps some block accessing methods (like `BlockByNumber` and `BlockByHash`). It also wraps raw RPC calls to create a golang interface for the Engine API (`newBlock`, `forkchoiceUpdated`, `getPayload`).

This PR also pulls out the ethclient initialization loop into a standalone function.

### CODE REVIEW

**Comment by norswap** [comment]
Note: https://github.com/ethereum-optimism/optimistic-specs/pull/260 is squashed as the first commit of this PR.

**Comment by norswap** [comment]
As for the new organization of interfaces, I think the litmus test would be a good written explanation of what all of these interfaces are, relate to one another, are created/held/passed.

**Comment by trianglesphere** [comment]
> What's the relationship between #260 and this?

#260 is a follow on targeting more of the L1 sources + sync start algorithm

**Comment by trianglesphere** [comment]
> As for the new organization of interfaces, I think the litmus test would be a good written explanation of what all of these interfaces are, relate to one another, are created/held/passed.

The `l2.Source` is the Ethereum client wrapper for the L2 execution engine. It directly wraps some block accessing methods (like `BlockByNumber` and `BlockByHash`). It also wraps raw RPC calls to create a golang interface for the Engine API (`newBlock`, `forkchoiceUpdated`, `getPayload`).


For the follow on:
This flattens the dependency hierarchy. Now there are two `sources` - one for L1 and one for L2 - that implement all access to L1/L2 data/instructions. It implements methods like the `BlockReference` directly in the L1/L2 source. The chainsource concept has been removed.

At sites that depend on L1/L2 data, they define an interface with the subset of method that they need, but the `l1.Source` and the `l2.Source` fulfill every required method (except for `FindSyncStart` but that requires both L1 and L2 data).

File: opnode/rollup/derive/payload_attributes.go [hunk 1]
```diff
@@ -236,91 +224,68 @@ func FilterBatches(config *rollup.Config, epoch rollup.Epoch, minL2Time uint64,
...
-			Timestamp:             l2.Uint64Quantity(batch.Timestamp),
-			Random:                randomnessSeed,
-			SuggestedFeeRecipient: config.FeeRecipientAddress,
-			Transactions:          txns,
-		}
+// SortedAndPreparedBatches turns a collection of batches to the input batches for a series of blocks
+func SortedAndPreparedBatches(batches []*BatchData, epoch, blockTime, minL2Time, maxL2Time uint64) []*BatchData {
+	m := make(map[uint64]*BatchData)
+	for _, b := range batches {
+		b := b // not sure if copy is required, but go does not make this easy
```
Comment (protolambda) [inline opnode/rollup/derive/payload_attributes.go:254]: Should not be required. We're only dealing with a pointer here, we already copy the pointer when we set `m[b.BatchV1.Timestamp]` below.

File: opnode/rollup/driver/step.go [hunk 2]
```diff
@@ -137,52 +159,73 @@ func (d *outputImpl) step(ctx context.Context, l2Head eth.BlockID, l2Finalized e
...
+	if updateSafe {
+		fc.SafeBlockHash = payload.BlockHash
+	}
+	if updateUnsafe {
+		fc.HeadBlockHash = payload.BlockHash
+	}
+	_, err = d.l2.ForkchoiceUpdate(ctx, &fc, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to persist execution payload: %v", err)
+		return nil, fmt.Errorf("Could not make the new L2 block canonical via forkchoice: %w", err)
```
Comment (protolambda) [inline opnode/rollup/driver/step.go:228]: nit: don't start error messages with a capital, Go style rules: https://github.com/golang/go/wiki/CodeReviewComments#error-strings  (I know, probably the most subjective thing in the whole guide, but my IDE will complain about it every time I open the file)

**Comment by protolambda** [comment]
Two minor things, generally looks good, and #260 addresses the API design, so I'm going to ignore that part here. Also one small merge conflict. Will review/approve when PR is updated

File: opnode/rollup/driver/step.go (hunk 2 above)
Comment (trianglesphere) [reply opnode/rollup/driver/step.go:228]: I didn't realize this. Will fix. Maybe we should add a lint for this

File: opnode/rollup/derive/payload_attributes.go (hunk 1 above)
Comment (trianglesphere) [reply opnode/rollup/derive/payload_attributes.go:254]: Right yes. Double checked and the problem is only when taking the address of the loop variable.

**Review by protolambda — APPROVED** [review]
LGTM

**Pull request merged**